### ISSUE TITLE
Autowiring for scalar types

### ISSUE THREAD

**Issue by zakirullin** [laminas/laminas-servicemanager#66]
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | no
| BC Break    | no

#### Summary

We had an enormous number of handcrafted factories in past. The fact is, once you need a scalar - you're in trouble, because `ReflecitionBasedAbstractFactory` has nothing to offer to you.
So we came up with our own solution - `AutowiringBasedFactory`, and it works well so far.

The only thing we should do to wire our scalars up:
```php
return function(array $config) {
   return [
      GetAppVersionHandler::class => [
          'lastCommitSha' => $config['lastCommitSha']->getString(),
      ],
      ...
];
```

That's it, just a few lines of code in `autowiring.php` file. And of course, you only need to provide problematic parameters, the rest will be handled for you as before.

Alongside with that we have a console command with following compilation-phase (no runtime is needed) checks:
1) Circular references
2) Compound/Scalar types match
3) Overall dependencies' config correctness 
4) Overall container consistency

With all that being said, is there any chance of our PR being merged? Given that it will satisfy all the requirements.

**Comment by zakirullin** [comment]
@weierophinney  @Ocramius  @boesing @geerteltink Hi there. Can you please have a look? 
It's fully BC, and  it adds some really useful possibilities, whereas giving prior-runtime container validation checks.

**Comment by Xerkus** [comment]
Just passing comment: this sounds more like a laminas-di territory: https://docs.laminas.dev/laminas-di/cookbook/aot-guide/

**Comment by zakirullin** [comment]
@Xerkus Thanks! As far as I can see, it's not widely used. And it definitely adds some complexity.
Whereas the solution above is fairly simple (5 lines of code above `ReflectionBasedAbstractionFactory.php`). Plus a separate container verification script. 
It doesn't have code generated factories though. We did some research and it seems like there's not much difference between Reflection-based approach and code generated factories. It was about 3ms which makes no sense (on fairly HUGE codebase).
Not to mention overall complexity & testability when it comes to code generation.

Both solutions have its own pros and cons, but let the users choose the way they want it to be :)

**Comment by Ocramius** [comment]
Discussed briefly in chat: I also believe that if we go down this road, it is not through `laminas-servicemanager`, but through `laminas-di`, with full AoT compilation of a PSR-11 container there.

Doing it here => too fragile already, and we don't have any AoT build mechanisms either.

**Comment by zakirullin** [comment]
Huh. Grasped the code of `laminas-di`. You were right, that's just **a perfect fit** for what I was trying to achieve :)
It works awesome when paired with `laminas-servicemanager`, few changes were required to a fairly huge codebase.
Great job!

The only features we're missing are so far:
- Container verification (without factories generation)
- Circular dependencies static check (now it's only at runtime)
- Container verification should also take in account `custom factories`/`ReflectionBasedAbstractFactory`

So, basically it's just a tool that uses `$config['dependencies']` to tell us whether it's all good or not (with certain degree of confidence).
It would benefit current users of `laminas-servicemanager` (no matter if used with `laminas-di`/`generated factories` or not).

**Comment by zakirullin** [comment]
https://github.com/laminas/laminas-servicemanager/pull/67
